src/richtext/utils/hastUtilToRichText.ts

```diff
@@ -0,0 +1,373 @@
+	RichTextHTMLMapSerializerShorthand | RichTextHTMLMapSerializerFunction
+>;
+
+const DEFAULT_SERIALIZER: RichTextHTMLMapSerializer = {
+	h1: "heading1",
+	h2: "heading2",
+	h3: "heading3",
+	h4: "heading4",
+	h5: "heading5",
+	h6: "heading6",
+	p: "paragraph",
+	pre: "preformatted",
+	strong: "strong",
+	em: "em",
+	li: ({ context }) =>
+		context.listType === "group-o-list-item" ? "o-list-item" : "list-item",
+	ul: "group-list-item",
+	ol: "group-o-list-item",
+	img: "image",
+	iframe: "embed",
+	a: "hyperlink",
+};
+
+const VFILE_RULE = "failed-to-serialize-node";
+const VFILE_SOURCE = "prismic";
+
+/**
+ * Configuration that determines the output of `hastUtilToRichText`.
+ */
+export type HastUtilToRichTextConfig = {
+	/**
+	 * An optional HTML to rich text serializer. Will be merged with the default
+	 * HTML to rich text serializer.
+	 */
+	serializer?: RichTextHTMLMapSerializer;
+
+	/**
+	 * Whether or not the text processed should be marked as right-to-left.
+	 *
+	 * @defaultValue `false`
+	 */
+	direction?: "ltr" | "rtl";
+
+	/**
+	 * The default node type to wrap the input with in case it only represents the
+	 * inner content of an element.
+	 *
+	 * @remarks
+	 * This is a niche option that allows you to serialize correctly an input such
+	 * as: `"lorem <strong>ipsum</strong> dolor sit amet"`, wrapping it inside a
+	 * node of the desired type.
+	 *
+	 * @defaultValue `"paragraph"`
+	 */
+	defaultWrapperNodeType?: RTTextNode["type"];
+};
+
+export const hastUtilToRichText = (
+	tree: Root | Element,
+	file: VFile,
+	config?: HastUtilToRichTextConfig,
+): RichTextField => {
+	const builder = new RichTextFieldBuilder();
+
+	// Merge the default serializer with the user-provided one.
+	const serializer = {
+		...DEFAULT_SERIALIZER,
+		...config?.serializer,
+	};
+
+	// Keep track of the last text node type to append text nodes to in
+	// case of an image or an embed node is present inside a paragraph.
+	let lastRTTextNodeType: RTTextNode["type"] =
+		config?.defaultWrapperNodeType ?? RichTextNodeType.paragraph;
+
+	// Keep track of the last list type to know whether we need to append
+	// `list-item` or `o-list-item` nodes.
+	let lastListType: "group-list-item" | "group-o-list-item" | null = null;
+
+	visit(tree, (node) => {
+		if (node.type === "element") {
+			// Transforms line break elements to line breaks
+			if (node.tagName === "br") {
+				try {
+					builder.appendText("\n");
+				} catch (error) {
+					// noop
+				}
+			}
+
+			// Resolves the serializer responsible for the current node.
+			let serializerOrShorthand:
+				| RichTextHTMLMapSerializerShorthand
+				| RichTextHTMLMapSerializerFunction;
+
+			// We give priority to CSS selectors over tag names.
+			if (node.matchesSerializer && node.matchesSerializer in serializer) {
+				serializerOrShorthand = serializer[node.matchesSerializer];
+			} else if (node.tagName in serializer) {
+				serializerOrShorthand = serializer[node.tagName];
+			} else {
+				return;
+			}
```

Comment:
Yes, basically I only populate nodes with `matchSerializer` when it's a CSS selector other than a tag name, because matching those is costly (we need to walk the whole tree once for **each** non-tag name CSS selectors). Matching tag names is cheap and should be preferred when possible like here (note the default serializer only uses tag names)

I'll see if I there's an opportunity when walking the tree previously to populate it by default with `matchSerializer` based on the tag name, otherwise this might be the best place to do it.

Reply:
Sounds good, that what I was expecting. Let's leave it as is. 🙂